ramips: add support for ZyXEL Keenetic Lite rev.B

robototechnic opened the pull request:
Device specification:
- SoC: RT5350
- CPU Frequency: 360 MHz
- Flash Chip: Macronix MX25L6406E (8192 KiB)
- RAM: Winbond W9825G6JH-6 (32768 KiB)
- 5x 10/100 Mbps Ethernet (4x LAN, 1x WAN)
- 1x external antenna
- UART (J1) header on PCB (57800 8n1)
- Wireless: SoC-intergated: 2.4GHz 802.11bgn
- USB: None
- 8x LED, 2x button

Flash instruction:
1. Configure PC with static IP 192.168.99.8/24 and start TFTP server.
2. Rename "openwrt-ramips-rt305x-zyxel_keenetic-lite_rev.b-squashfs-sysupgrade.bin"
   to "rt305x_firmware.bin" and place it in TFTP server directory.
3. Connect PC with one of LAN ports, press the reset button, power up
   the router and keep button pressed until power LED start blinking.
4. Router will download file from TFTP server, write it to flash and reboot.

Signed-off-by: robototechnic <senior.anonymous@ya.ru>

adrianschmutzler commented:
Hi, please add the specs and flashing instructions to the commit message. Despite, you need to use your real name for Signed off by.

adrianschmutzler commented on target/linux/ramips/dts/rt5350_zyxel_keenetic-lite_rev.b.dts:
```diff
@@ -0,0 +1,121 @@
+/dts-v1/;
```
Please add an SPDX license identifier.

robototechnic replied:
Added

adrianschmutzler commented on target/linux/ramips/dts/rt5350_zyxel_keenetic-lite_rev.b.dts:
```diff
@@ -0,0 +1,121 @@
+#include <dt-bindings/input/input.h>
+
+/ {
+	compatible = "zyxel,keenetic-lite_rev.b", "ralink,rt5350-soc";
```
I'd use "zyxel,keenetic-lite-b" here, and adjust everything else accordingly.

robototechnic replied:
renamed

adrianschmutzler commented on target/linux/ramips/dts/rt5350_zyxel_keenetic-lite_rev.b.dts:
```diff
@@ -0,0 +1,121 @@
+
+/ {
+	compatible = "zyxel,keenetic-lite_rev.b", "ralink,rt5350-soc";
+	model = "ZyXEL Keenetic Lite rev.B";
```
I'd use "Rev. B" instead of "rev.B"

robototechnic replied:
renamed

adrianschmutzler commented on target/linux/ramips/dts/rt5350_zyxel_keenetic-lite_rev.b.dts:
```diff
@@ -0,0 +1,121 @@
+		compatible = "gpio-leds";
+
+		led_status: power {
+			label = "zyxel:green:power";
```
While the keenetic start is an exception, it's common to use the model part of the device name here, so "keenetic-lite-b:" instead of "zyxel:" if you follow my suggestion above.

robototechnic replied:
fixed

adrianschmutzler commented on target/linux/ramips/dts/rt5350_zyxel_keenetic-lite_rev.b.dts:
```diff
@@ -0,0 +1,121 @@
+	leds {
+		compatible = "gpio-leds";
+
+		led_status: power {
```
"led_power: power {"

robototechnic replied:
fixed

adrianschmutzler commented on target/linux/ramips/dts/rt5350_zyxel_keenetic-lite_rev.b.dts:
```diff
@@ -0,0 +1,121 @@
+
+&state_default {
+	gpio {
+		groups = "i2c", "jtag", "uartf";
```
Do you need all of these?

robototechnic replied:
I do not need, but the chip supports these protocols. If necessary, I will remove them.

adrianschmutzler replied:
Actually, it's the other way around. See here for details: https://github.com/openwrt/openwrt/pull/2988#issuecomment-623867267

robototechnic replied:
Left only uartf

adrianschmutzler commented on target/linux/ramips/dts/rt5350_zyxel_keenetic-lite_rev.b.dts:
```diff
@@ -0,0 +1,121 @@
+	flash@0 {
+		compatible = "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <10000000>;
```
Can this be increased?

robototechnic replied:
I increased it to 40,000,000 and install the image, it works fine, I did not notice any changes in the speed of work.

adrianschmutzler replied:
This can be tested by reading a defined amount of data from flash (e.g. `dd if=/dev/mtdX of=/dev/null bs=2M count=1`)

Despite, it might also be a good idea to check the maximum speed in the flash specifications.

robototechnic replied:
I can not find the maximum flash speed in the specification. Searched [here](https://cdn.sparkfun.com/datasheets/Wireless/WiFi/RT5350.pdf) and [here](http://read.pudn.com/downloads622/ebook/2530552/Ralink_SDK_User_Manual_4.2.pdf). Experimentally found the maximum speed (60,000,000). With a further increase, the device does not boot.
At 10,000,000, the command `time dd if=/dev/mtdblock6 of=/dev/null bs=10M count=1` is executed in 8.3s
at 40,000,000 in 4.01s
at 60,000,000 in 3.5s

mans0n replied:
So you mean `spi-max-frequency = <60000000>;` worked? Interesting, I thought that frequency is too high for RT5350.
BTW you can check the exact hardware frequency by changing this line from dev_dbg to dev_info.
https://github.com/openwrt/openwrt/blob/master/target/linux/ramips/patches-4.14/0042-SPI-ralink-add-Ralink-SoC-spi-driver.patch#L234

robototechnic replied:
When I change this line `dmesg` shows the frequency that I set.
`root@OpenWrt:~# dmesg | grep spi`
`[    0.607320] spi spi0.0: force spi mode3`
`[    0.615666] m25p80 spi0.0: speed: 60000000/60000000, rate: 2, prescal: 0, loops: 130` #60000000/**60000000** - this speed is always the same as I set in the device tree file
`[    0.632101] m25p80 spi0.0: mx25l6405d (8192 Kbytes)`
`[    0.642048] 4 fixed-partitions partitions found on MTD device spi0.0`
`[    0.654791] Creating 4 MTD partitions on "spi0.0":`

mans0n replied:
Good to hear that. I've never seen a device with rate 2 working, but this device seems to work.

adrianschmutzler commented on target/linux/ramips/rt305x/base-files/etc/board.d/01_leds:
```diff
@@ -42,6 +42,9 @@ nexx,wt1520-4m|\
 skyline,sl-r7205|\
 tenda,w306r-v2|\
+zyxel,keenetic-lite_rev.b)
+	set_wifi_led "rt2800soc-phy0::radio"
```
This will break all devices above.
Despite, why does this deviate from keenetic-start?

robototechnic replied:
fixed

adrianschmutzler replied:
So, what's with my question? Why have you chanded the trigger now?

robototechnic replied:
The trigger remains the same (set_wifi_led "rt2800soc-phy0 :: radio"). Based on https://github.com/trexxet/openwrt-zyxel_keenetic/blob/master/0019-ramips-add-ZyXel-Keenetic-userspace.patch
Working fine, I uploaded a new image after making changes.

mans0n replied:
Could you check what `grep . /sys/class/leds/*/trigger` returns?

robototechnic replied:
`root@OpenWrt:~# grep .  /sys/class/leds/*/trigger`
`/sys/class/leds/keenetic-lite-b:green:internet/trigger:[none] timer default-on netdev heartbeat phy0rx phy0tx phy0assoc phy0radio phy0tpt`
`/sys/class/leds/keenetic-lite-b:green:power/trigger:[none] timer default-on netdev heartbeat phy0rx phy0tx phy0assoc phy0radio phy0tpt`
`/sys/class/leds/keenetic-lite-b:green:wlan/trigger:[none] timer default-on netdev heartbeat phy0rx phy0tx phy0assoc phy0radio phy0tpt`
`/sys/class/leds/keenetic-lite-b:green:wps/trigger:[none] timer default-on netdev heartbeat phy0rx phy0tx phy0assoc phy0radio phy0tpt`
`/sys/class/leds/rt2800soc-phy0::assoc/trigger:[none] timer default-on netdev heartbeat phy0rx phy0tx phy0assoc phy0radio phy0tpt`
`/sys/class/leds/rt2800soc-phy0::quality/trigger:[none] timer default-on netdev heartbeat phy0rx phy0tx phy0assoc phy0radio phy0tpt`
`/sys/class/leds/rt2800soc-phy0::radio/trigger:[none] timer default-on netdev heartbeat phy0rx phy0tx phy0assoc phy0radio phy0tpt
`

mans0n replied:
So there's no `rt2800pci-phy0::radio` LED on your device, and none of your LEDs is using phy0* trigger. Does any LED turn on whey you enable wlan?

robototechnic replied:
I so sorry for this. My programmer skill is very low. Moved the router to the `set_wifi_led trigger "rt2800soc-phy0 :: radio"` trigger now. Uploaded new image, work fine.
Interestingly, with the `rt2800pci-phy0::radio` trigger, the wifi LED lights up when the wlan is turned on.

mans0n replied:
If the wifi led uses WLAN_LED_N pin, then you don't need any LED setting, it will automatically turn on when you enable wifi.
You can turn on a LED by `echo 1 > /sys/class/leds/<LED>/brightness`
and turn it off by `echo 0 > /sys/class/leds/<LED>/brightness`.
Could you check if all LEDs turn on/off correctly with this command?

robototechnic replied:
work:
`echo 1 > /sys/class/leds/keenetic-lite-b:green:power/brightness`
`echo 1 > /sys/class/leds/keenetic-lite-b:green:wps/brightness`
`echo 1 > /sys/class/leds/rt2800soc-phy0::radio/brightness`

not work:
`echo 1 > /sys/class/leds/keenetic-lite-b:green:wlan/brightness`
`echo 1 > /sys/class/leds/keenetic-lite-b:green:internet/brightness`
`echo 1 > /sys/class/leds/rt2800soc-phy0::assoc/brightness`
`echo 1 > /sys/class/leds/rt2800soc-phy0::quality/brightness`

mans0n replied:
So it seems wlan and internet LED do not exist, please drop the nodes in dts. And probably you won't need the 01_leds setting as well.

robototechnic replied:
Removed wlan and internet LEDs. Returned the `01_leds` file to its original state.

adrianschmutzler commented on target/linux/ramips/rt305x/base-files/etc/board.d/02_network:
```diff
@@ -201,6 +202,7 @@ ramips_setup_macs()
 	poray,ip2202|\
 	teltonika,rut5xx|\
 	unbranded,xdx-rn502j|\
+	zyxel,keenetic-lite_rev.b|\
```
Is there an appropriate address in 0x2e?
Does the device have a label with a MAC address on it and if yes, to which interface does it correspond?

robototechnic replied:
deleted this line

adrianschmutzler replied:
Why? And what's with my question?

mans0n replied:
Does `rt2800pci-phy0::radio` work? I've never seen it on RT5350 before.

robototechnic replied:
Unfortunately I do not know the answer to this question, I just adapt the old patch for the new version of openwrt
https://github.com/trexxet/openwrt-zyxel_keenetic/blob/master/0019-ramips-add-ZyXel-Keenetic-userspace.patch

adrianschmutzler replied:
Just run the following and put the result here:
```
. /lib/functions/system.sh
. /lib/functions.sh
mtd_get_mac_binary factory 0x4
mtd_get_mac_binary factory 0x28
mtd_get_mac_binary factory 0x2e
mtd_get_mac_binary factory 0x8004
mtd_get_mac_binary factory 0xe000
mtd_get_mac_binary factory 0xe006
```

Only the last byte of the addresses is relevant, so you might just hide the rest of the returned addresses if you want to, e.g.
xx:xx:xx:xx:xx:34

And please do the same for the addresses on the device's outside label, if it has one.

robototechnic replied:
`root@OpenWrt:~# . /lib/functions/system.sh`
`root@OpenWrt:~# . /lib/functions.sh`
`root@OpenWrt:~# mtd_get_mac_binary factory 0x4`
`xx:xx:xx:xx:xx:5f`
`root@OpenWrt:~# mtd_get_mac_binary factory 0x28`
`xx:xx:xx:xx:xx:60`
`root@OpenWrt:~# mtd_get_mac_binary factory 0x2e`
`ff:ff:ff:ff:ff:ff`
`root@OpenWrt:~# mtd_get_mac_binary factory 0x8004`
`ff:ff:ff:ff:ff:ff`
`root@OpenWrt:~# mtd_get_mac_binary factory 0xe000`
`ff:ff:ff:ff:ff:ff`
`root@OpenWrt:~# mtd_get_mac_binary factory 0xe006`
`ff:ff:ff:ff:ff:ff`
>And please do the same for the addresses on the device's outside label, if it has one.

Sorry, the case of the router has long been thrown away.

mans0n replied:
[Googling](https://60.img.avito.st/1280x960/8128330760.jpg) shows this model has WAN MAC address on its label

robototechnic replied:
Yes it does, but I only have a bare board. The case on which the label was located I threw away a long time ago. If this is so important, I can ask the people on the forum to execute these commands and send me a response and the photo of label.

adrianschmutzler commented on target/linux/ramips/image/rt305x.mk:
```diff
@@ -1192,6 +1192,14 @@ define Device/zyxel_keenetic
+  SOC := rt5350
+  IMAGE_SIZE := 7872k
+  DEVICE_VENDOR := ZyXEL
+  DEVICE_MODEL := Keenetic Lite rev.B
```
I'd go for
DEVICE_MODEL := Keenetic Lite
DEVICE_VARIANT := B

robototechnic replied:
fixed

adrianschmutzler commented on target/linux/ramips/image/rt305x.mk:
```diff
@@ -1192,6 +1192,14 @@ define Device/zyxel_keenetic
 endef
 TARGET_DEVICES += zyxel_keenetic
 
+define Device/zyxel_keenetic-lite_rev.b
```
Same as above, I'd go for zyxel_keenetic-lite-b here.

robototechnic replied:
renamed

robototechnic commented:
> Hi, please add the specs and flashing instructions to the commit message. Despite, you need to use your real name for Signed off by.

Thank you for spending time on my PR, it my first time.

adrianschmutzler commented:
Please squash your changes into a single commit (git rebase -i ...) and then force-push them here again.

Commit:
ramips: add support for ZyXEL Keenetic Lite Rev.B

Device specification:

SoC: RT5350
CPU Frequency: 360 MHz
Flash Chip: Macronix MX25L6406E (8192 KiB)
RAM: Winbond W9825G6JH-6 (32768 KiB)
5x 10/100 Mbps Ethernet (4x LAN, 1x WAN)
1x external antenna
UART (J1) header on PCB (57800 8n1)
Wireless: SoC-intergated: 2.4GHz 802.11bgn
USB: None
8x LED, 2x button

Flash instruction:

Configure PC with static IP 192.168.99.8/24 and start TFTP server.
Rename "openwrt-ramips-rt305x-zyxel_keenetic-lite-b-squashfs-sysupgrade.bin"
to "rt305x_firmware.bin" and place it in TFTP server directory.
Connect PC with one of LAN ports, press the reset button, power up
the router and keep button pressed until power LED start blinking.
Router will download file from TFTP server, write it to flash and reboot.

Signed-off-by: Sergei Burakov <senior.anonymous@ya.ru>

mans0n commented on target/linux/ramips/dts/rt5350_zyxel_keenetic-lite-b.dts:
```diff
@@ -0,0 +1,122 @@
+
+		led_power: power {
+			label = "keenetic-lite-b:green:power";
+			gpios = <&gpio0 9 GPIO_ACTIVE_LOW>;
```
I'd be happy if you rearrange led/key nodes by gpio pin number.

robototechnic replied:
Moved it.

mans0n commented on target/linux/ramips/dts/rt5350_zyxel_keenetic-lite-b.dts:
```diff
@@ -0,0 +1,122 @@
+	};
+
+	keys {
+		compatible = "gpio-keys-polled";
```
Try using "gpio-keys" and dropping poll-interval.

robototechnic replied:
Change it. Tested, work fine.

mans0n commented:
Could you remove leading whitespaces in the commit message to fit the lines into 80-characters limit?

And note that this is 32M device so its support might be limited in the furure.

Not so important but have you also tried kernel 5.4? Testing new kernel version is welcome.

robototechnic commented:
> Could you remove leading whitespaces in the commit message to fit the lines into 80-characters limit?
> 
> And note that this is 32M device so its support might be limited in the future.
> 
> Not so important but have you also tried kernel 5.4? Testing new kernel version is welcome.
1) Fixed
2) Add note to comment?
3) I’ll try to test, tell me please where to see how to compile an image with a new core?

adrianschmutzler commented:
>> And note that this is 32M device so its support might be limited in the future.
>
> Add note to comment?

RAM size is already mentioned in the commit message, that's fine. The original was just FYI.

robototechnic commented:
> RAM size is already mentioned in the commit message, that's fine. The original was just FYI.
Yes, I know about that. I read [this](https://openwrt.org/supported_devices/432_warning) page when I first time compiled the image. Thank you.

mans0n commented on target/linux/ramips/dts/rt5350_zyxel_keenetic-lite-b.dts:
```diff
@@ -0,0 +1,121 @@
+};
+
+&wmac {
+	status = "okay";
```
wmac is not disabled in dtsi so you don't need to enable it.

robototechnic replied:
removed `status = "okay";`

mans0n commented:
> I’ll try to test, tell me please where to see how to compile an image with a new core?

You can try 5.4 by selecting "Global build settings" -> "Use the testing kernel version" in menuconfig.

robototechnic commented:
> You can try 5.4 by selecting "Global build settings" -> "Use the testing kernel version" in menuconfig.

It seems to be working fine.
![2020-06-10_11-40-37](https://user-images.githubusercontent.com/34556074/84247586-fc5cfb80-ab10-11ea-892a-a9c21bf15eee.png)

robototechnic commented:
Need any more changes?

robototechnic commented:
Hello!
Do I need to do something else?

ynezz commented:
>  Do I need to do something else?

Nope, just be patient :)

ynezz commented:
Thanks! Pulled into my staging tree at https://git.openwrt.org/openwrt/staging/ynezz.git
